constructor/header.sh

```diff
@@ -1,89 +1,125 @@
-#!/bin/bash
+#!/bin/sh
 #
 # NAME:  __NAME__
 # VER:   __VERSION__
 # PLAT:  __PLAT__
 # BYTES: @SIZE_BYTES@
 # LINES: @LINES@
 # MD5:   __MD5__
-
 #if osx
 unset DYLD_LIBRARY_PATH
 #else
 export OLD_LD_LIBRARY_PATH=$LD_LIBRARY_PATH
 unset LD_LIBRARY_PATH
 #endif
 
-echo "$0" | grep '\.sh$' >/dev/null
-if (( $? )); then
+if ! echo "$0" | grep '\.sh$' >/dev/null; then
     echo 'Please run using "bash" or "sh", but not "." or "source"' >&2
     return 1
 fi
 
-THIS_DIR=$(cd $(dirname $0); pwd)
-THIS_FILE=$(basename $0)
+# Determine RUNNING_SHELL; if SHELL is non-zero use that.
+if [ -n "$SHELL" ]; then
+    RUNNING_SHELL="$SHELL"
+else
+    if [ "$(uname)" = "Darwin" ]; then
+        RUNNING_SHELL=/bin/bash
+    else
+        RUNNING_SHELL=$(ps -p $$ -oargs=)
```

Comment:
How robust is part of the branch?  AFAICT most shells *do* define `$SHELL`, so how well is this logic understood/tested/etc.?
Also, why a special case for Darwin?  Is this to handle non-OSX Darwin, or...?